crates/types/src/storage.rs

```diff
@@ -395,6 +399,126 @@ pub fn decode_from_flexbuffers<T: DeserializeOwned, B: Buf>(
     }
 }
 
+/// [`ArcVec`] mainly used by `message` types to improve
+/// cloning of messages.
+///
+/// It can replace [`Vec<T>`] most of the time in all structures
+/// that need to be serialized over the wire.
+///
+/// Internally it keeps the data inside an [`Arc<[T]>`]
+#[derive(Debug)]
+pub struct ArcVec<T> {
+    inner: Arc<[T]>,
```

Comment:
Is there a benefit over `Arc<Vec<T>>`? If I understand the API correctly, then converting from `Vec<T>` into an `Arc<[T]>` requires an extra allocation whereas the former adds another level of pointer indirection.

Reply:
The only reason that I used `Arc<[T]>` instead of `Arc<Vec<T>>` is because `Arc<[T]>` is very common type in Bifrost.

In Bifrost the `Loglet::enqueue_batch` method accepts an `Arc<[Record]>` as a param, but when we actually need to send this over the wire (in case of replicated loglet client, or to log servers), the message data types are using `Vec<Record>` type. This requires copying all records to build, which can be quite heavy in case we sending the same set of records to multiple log servers for example.

In other wards, the most common usage patter is to build an ArcVec from an `Arc<[Record]>` not from Vec.

That being said, I should actually `Arc<[T]>::from(vec)` instead of `from_iter` to build the ArcVec from a Vec (in case of deseralization) to avoid one extra allocation. as per [here](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-From%3CVec%3CT,+A%3E%3E-for-Arc%3C%5BT%5D,+A%3E)

Reply:
Thanks for the explanation.

I think that `Arc<[T]>::from(vec)` will also require one additional allocation.

Once this becomes a problem and measurable, we can revisit it.